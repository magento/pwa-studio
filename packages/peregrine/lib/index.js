import * as RestApi from './RestApi';
import * as Util from './util';

// hooks
export { useEventListener } from './hooks/useEventListener';
export { useCarousel } from './hooks/useCarousel';
export { useDropdown } from './hooks/useDropdown';
export { usePagination } from './hooks/usePagination';
export { useRestApi } from './hooks/useRestApi';
export { useRestResponse } from './hooks/useRestResponse';
export { useScrollLock } from './hooks/useScrollLock';
export { useSearchParam } from './hooks/useSearchParam';
export { useSort } from './hooks/useSort';
export { useTypePolicies } from './hooks/useTypePolicies';

export {
    WindowSizeContextProvider,
    useWindowSize
} from './hooks/useWindowSize';
export { getToastId, useToasts, ToastContextProvider } from './Toasts';

// store
export { enhancer, reducers } from './store';

// components
export { default as ContainerChild } from './ContainerChild';
export { default as List, Items, Item } from './List';
export { default as Price } from './Price';
export { default as Router } from './Router';
export {
    default as PeregrineContextProvider
} from './PeregrineContextProvider';

// misc
export { RestApi };
export { Util };
export { default as createTestInstance } from './util/createTestInstance';
