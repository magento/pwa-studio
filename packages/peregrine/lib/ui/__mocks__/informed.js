import React from 'react';

const informed = jest.requireActual('informed');
const {
    Text,
    Input,
    Radio,
    TextArea,
    Select,
    Option,
    Checkbox,
    RadioGroup,
    BasicText,
    BasicRadio,
    BasicRadioGroup,
    BasicTextArea,
    BasicSelect,
    BasicCheckbox,
    Scope,
    ArrayField,
    asField,
    ...rest
} = informed;

/**
 * HoC to wrap informed form field components. This HoC modifies
 * the ID prop of the given component and replaces it with the
 * value of the field prop if available. If not it replaces it with
 * a sample value `SAMPLE_ID`.
 *
 * We are doing this because the new version of informed added a feature
 * to use UUID as the value of the ID prop if one is not provided.
 * This is breaking all our snapshot tests because a new ID is created
 * every time the component renders.
 *
 * https://github.com/joepuzzo/informed/blob/master/src/utils.js#L123
 *
 * @param {ReactComponent} Component component to wrap
 *
 * @returns {ReactComponent}
 */
const wrapWithUuidPropCheck = Component => props => {
    const newProps = { ...props };
    const uuid4Regex = new RegExp(
        /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i
    );
    const hasUuidIdProp =
        props.hasOwnProperty('id') && uuid4Regex.test(props.id);

    if (!props.hasOwnProperty('id') || hasUuidIdProp) {
        if (props.hasOwnProperty('field')) {
            newProps.id = props.field;
        } else {
            newProps.id = 'SAMPLE-ID';
        }
    }

    return <Component {...newProps} />;
};

module.exports = {
    ...rest,
    Text: wrapWithUuidPropCheck(Text),
    Input: wrapWithUuidPropCheck(Input),
    Radio: wrapWithUuidPropCheck(Radio),
    TextArea: wrapWithUuidPropCheck(TextArea),
    Select: wrapWithUuidPropCheck(Select),
    Option: wrapWithUuidPropCheck(Option),
    Checkbox: wrapWithUuidPropCheck(Checkbox),
    RadioGroup: wrapWithUuidPropCheck(RadioGroup),
    BasicText: wrapWithUuidPropCheck(BasicText),
    BasicRadio: wrapWithUuidPropCheck(BasicRadio),
    BasicRadioGroup: wrapWithUuidPropCheck(BasicRadioGroup),
    BasicTextArea: wrapWithUuidPropCheck(BasicTextArea),
    BasicSelect: wrapWithUuidPropCheck(BasicSelect),
    BasicCheckbox: wrapWithUuidPropCheck(BasicCheckbox),
    Scope: wrapWithUuidPropCheck(Scope),
    ArrayField: wrapWithUuidPropCheck(ArrayField),
    asField: Component => wrapWithUuidPropCheck(asField(Component))
};
