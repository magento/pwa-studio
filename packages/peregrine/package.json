{
  "name": "@magento/peregrine",
  "version": "12.1.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "clean": " "
  },
  "repository": "github:magento/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento/pwa-studio/tree/master/packages/peregrine#readme",
  "dependencies": {
    "fast-glob": "~3.2.4"
  },
  "devDependencies": {
    "@apollo/client": "~3.4.0",
    "@magento/eslint-config": "~1.5.0",
    "@testing-library/react-hooks": "~5.0.3",
    "intl": "~1.2.5",
    "intl-locales-supported": "~1.8.12",
    "react": "~17.0.1",
    "react-dom": "~17.0.1",
    "react-intl": "~5.20.0",
    "react-redux": "~7.2.2",
    "react-router-dom": "~5.2.0",
    "react-test-renderer": "~17.0.1",
    "redux": "~4.0.4",
    "redux-actions": "~2.6.5",
    "wait-for-expect": "~1.2.0"
  },
  "peerDependencies": {
    "@apollo/client": "~3.4.0",
    "@babel/runtime": "~7.15.3",
    "informed": "~3.29.0",
    "react": "~17.0.1",
    "react-dom": "~17.0.1",
    "react-intl": "~5.20.0",
    "react-redux": "~7.2.2",
    "react-router-dom": "~5.2.0",
    "redux": "~4.0.4",
    "redux-actions": "~2.6.5"
  },
  "engines": {
    "node": ">=14.x",
    "yarn": ">=1.12.0"
  },
  "pwa-studio": {
    "targets": {
      "declare": "./lib/targets/peregrine-declare",
      "intercept": "./lib/targets/peregrine-intercept"
    }
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
