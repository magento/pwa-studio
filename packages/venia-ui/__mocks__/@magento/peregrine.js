import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
const useTypePolicies = jest.fn(peregrine.useTypePolicies);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useEventListener,
    useDropdown,
    useSearchParam,
    useWindowSize,
    useToasts,
    useTypePolicies
};
