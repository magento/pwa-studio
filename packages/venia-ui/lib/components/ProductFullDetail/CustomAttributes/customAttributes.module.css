.root {
}

.list {
    composes: list-disc from global;
    composes: mb-4 from global;
    composes: mt-4 from global;
    composes: pl-6 from global;
}
