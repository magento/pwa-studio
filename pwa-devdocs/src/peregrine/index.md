---
title: Peregrine
adobeio: /guides/packages/peregrine/
---

The Peregrine package is a collection of functions that act as the brains of your visual components.
They provide pieces of logic for UI components, such as the ones provided by the [Venia][] library.
Use or remix these functions and components to create a unique Magento PWA storefront.

## Custom React hooks

Many of the functions provided by Peregrine are [custom React hooks][].
This lets them maintain an internal state without relying on an external library, such as Redux.

Peregrine hooks are designed to be flexible, and non-opinionated about UI.
They contain code for providing data or behavior logic and do not render content themselves.
Rendering content is left up to UI components.

Separating logic and presentation code gives developers more flexibility on how to use PWA Studio components with their own custom code.
A developer may choose to use a Venia feature that uses certain Peregrine hooks with minor visual modifications, or
they can use those same Peregrine hooks to develop their own feature with a different UI.

For more information about custom hooks, see the React documentation for [Building Your Own Hooks][].

### Return signatures

The return signatures of Peregrine hooks vary and is dependent on their purpose.

Some return an array with *state* and *api* objects, which follow the same pattern as [`useState()`][] and [`useReducer()`].
This lets you specify the variable names of the provided objects when you [destructure][] the array.

Other Peregrine hooks return a single object.

Use the reference docs on this site or in the [JSDoc][] blocks in the source code learn the API for each hook.

## JSDoc blocks

Most of the reference docs in this section are generated from [JSDoc][] blocks.
The currently published docs reflect what is available from the most recent release, but
the JSDoc blocks in the source are kept up to date for developers who want to work with unreleased code.

## Source

Visit the [`peregrine`][] package to view the source or contribute to this project.

[venia]: {%link venia-pwa-concept/index.md %}

[`peregrine`]: https://github.com/magento/pwa-studio/tree/master/packages/peregrine
[custom react hooks]: https://reactjs.org/docs/hooks-custom.html
[destructure]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
[jsdoc]: https://devdocs.io/jsdoc/
[`usestate()`]: https://reactjs.org/docs/hooks-reference.html#usestate
[`usereducer()`]: https://reactjs.org/docs/hooks-reference.html#usereducer
[building your own hooks]: https://reactjs.org/docs/hooks-custom.html
