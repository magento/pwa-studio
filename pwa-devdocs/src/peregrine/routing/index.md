---
title: Routing in PWA Studio
adobeio: /guides/general-concepts/routing/
---

For web applications, routing is the process of mapping a URL to specific page resources.

In a multi-page application, routing is performed on the server side.
Every URL request fetches new HTML from the server and the browser loads the entire page.
This approach is inefficient because the same assets get loaded every time site navigation occurs.

For a single-page application (SPA), such as a progressive web app, routing is performed on the client side.
Single-page applications do not reload the browser during navigation.
Instead, the application uses the URL to fetch smaller pieces of data from the server and updates specific items on the page.

## Routing for PWA Studio storefronts

PWA Studio provides tools that support both server-side and client-side routing.

### Server-side routing

Server-side routing is accomplished using the [UPWARD][] configuration file.
Since the configuration file defines how the server responds to requests,
you can specify a different template to render each page type, such as a CMS page or a product details page.

Early versions of the Venia storefront used this approach, but
in the current version, every page request now returns the same HTML with the application shell.
The application decides how the page should render based on the request.

### Client-side routing

Client-side routing happens inside the storefront application.
When a user navigates inside the application, it updates the relevant pieces instead of refreshing the entire page to update content.

Since Venia is a single-page application, it uses client-side routing for internal navigation.

## How routing works in Venia

This section goes over the routing flow implemented in Venia.
It is the default workflow for all new projects created using the scaffolding tool,
but it is not the only possible workflow for routing.

### Initial request

Venia's UPWARD server handles the initial request to the storefront application.
Its `upward.yml` configuration tells the server to return an `index.html` page created at build time by Webpack, via the [HtmlWebpackPlugin][].
The content of this file is the same for all page types.

After the browser loads the application shell, routing is handled client-side by React Router.

### Routes component

Used inside the [App component][], the [Routes component][] provides the switch logic for deciding which component to use to render the main content for the current route.

Routes in a PWA do not have to correspond to routes in the old Magento storefront. Venia has routes not defined in the Magento backend, such as _Create account_ or full page _Checkout_.
These routes are assigned components which render content for those pages in the Routes component.

### MagentoRoute component

The [MagentoRoute component][] handles the routes that are Magento-specific, such as a product or category page.
It uses its [Peregrine talon][] counterpart to determine which component to display.

### useMagentoRoute() talon

The [`useMagentoRoute()`][] talon returns the correct component for a page type. It uses the [`getRouteComponent()`][] helper function to get the **root component** associated with a page type.
It uses the `resolveUnknownRoute()` function to determine the page type for a route and retrieves the root component associated with that type using a global `fetchRootComponent` function.
The `fetchRootComponent` function is autogenerated at build time by the Buildpack RootComponentsPlugin, which searches for RootComponents in the project and builds a manifest.

### resolveUnknownRoute()

The [`resolveUnknownRoute()`][] function is a utility function for fetching page type information from the backend Magento server using a GraphQL query.
The `getRouteComponent()` function uses the information from this query to get the correct root component from an object that maps page types to root components.

[upward]: {%link venia-pwa-concept/features/modular-components/index.md %}
[peregrine talon]: {%link peregrine/talons/index.md %}

[app component]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/App/app.js
[routes component]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/Routes/routes.js
[magentoroute component]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js
[`usemagentoroute()`]: https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js
[`getroutecomponent()`]: https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/MagentoRoute/getRouteComponent.js
[`resolveunknownroute()`]: https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/Router/resolveUnknownRoute.js
[htmlwebpackplugin]: https://webpack.js.org/plugins/html-webpack-plugin/
