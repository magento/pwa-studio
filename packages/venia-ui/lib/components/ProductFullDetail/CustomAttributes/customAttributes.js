import React, { useMemo } from 'react';
import { array, shape, string } from 'prop-types';

import { useStyle } from '@magento/venia-ui/lib/classify';

import AttributeType from './AttributeType';
import defaultClasses from './customAttributes.module.css';

export const IS_VISIBLE_ON_FRONT = 'PRODUCT_DETAILS_PAGE';

const CustomAttributes = props => {
    const { customAttributes, showLabels } = props;
    const classes = useStyle(defaultClasses, props.classes);

    const list = useMemo(
        () =>
            customAttributes.reduce((previousAttribute, currentAttribute) => {
                const usedInComponents =
                    currentAttribute.attribute_metadata?.used_in_components ||
                    [];
                // Visible on front attributes only
                if (usedInComponents.includes(IS_VISIBLE_ON_FRONT)) {
                    const attributeContent = (
                        <li
                            key={currentAttribute.attribute_metadata.uid}
                            className={classes.listItem}
                        >
                            <AttributeType
                                showLabels={showLabels}
                                data={currentAttribute}
                            />
                        </li>
                    );

                    previousAttribute.push(attributeContent);
                }

                return previousAttribute;
            }, []),
        [classes, customAttributes, showLabels]
    );

    if (list.length === 0) {
        return null;
    }

    return (
        <div className={classes.root}>
            <ul className={classes.list}>{list}</ul>
        </div>
    );
};

CustomAttributes.propTypes = {
    classes: shape({
        root: string,
        title: string,
        list: string,
        listItem: string
    }),
    customAttributes: array.isRequired
};

export default CustomAttributes;
