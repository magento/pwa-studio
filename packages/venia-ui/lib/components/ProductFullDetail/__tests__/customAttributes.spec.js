import React from 'react';
import { act } from 'react-test-renderer';
import { Form } from 'informed';

import { createTestInstance } from '@magento/peregrine';

import CustomAttributes from '../customAttributes';

let inputProps = {};

const Component = () => {
    return <CustomAttributes {...inputProps} />;
};

const givenDefaultValues = () => {
    inputProps = {
        customAttributes: []
    };
};

const givenCustomAttributes = () => {
    inputProps = {
        customAttributes: []
    };
};

describe('#CustomAttributes', () => {
    beforeEach(() => {
        givenDefaultValues();
    });

    it('renders without custom attributes', () => {
        const tree = createTestInstance(<Component />);

        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('renders with custom attributes not visible on front', () => {
        inputProps = {
            customAttributes: [
                {
                    attribute_metadata: {
                        is_visible_on_front: false,
                        sort_order: 2
                    }
                },
                {
                    attribute_metadata: {
                        is_visible_on_front: false,
                        sort_order: 1
                    }
                }
            ]
        };

        const tree = createTestInstance(<Component />);

        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('renders with custom attributes visible on front', () => {
        inputProps = {
            customAttributes: [
                {
                    attribute_metadata: {
                        is_visible_on_front: true,
                        label: 'Attribute 2',
                        sort_order: 2,
                        uid: 'uid-2'
                    },
                    entered_attribute_value: {
                        value: 'Option 3'
                    }
                },
                {
                    attribute_metadata: {
                        is_visible_on_front: true,
                        label: 'Attribute 1',
                        sort_order: 1,
                        uid: 'uid-1'
                    },
                    selected_attribute_options: {
                        attribute_option: [
                            {
                                label: 'Option 1'
                            },
                            {
                                label: 'Option 2'
                            }
                        ]
                    }
                },
                {
                    attribute_metadata: {
                        is_visible_on_front: true,
                        label: 'Attribute 3',
                        sort_order: 3,
                        uid: 'uid-3'
                    },
                    selected_attribute_options: {
                        attribute_option: []
                    }
                },
                {
                    attribute_metadata: {
                        is_visible_on_front: false,
                        sort_order: 4
                    }
                }
            ]
        };

        const tree = createTestInstance(<Component />);

        expect(tree.toJSON()).toMatchSnapshot();
    });
});
