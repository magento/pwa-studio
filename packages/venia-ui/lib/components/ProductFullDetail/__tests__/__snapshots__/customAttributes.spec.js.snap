// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CustomAttributes renders with custom attributes not visible on front 1`] = `null`;

exports[`#CustomAttributes renders with custom attributes visible on front 1`] = `
<div>
  <p>
    <mock-FormattedMessage
      defaultMessage="Details"
      id="customAttributes.title"
    />
  </p>
  <ul>
    <li>
      <span>
        Attribute 1
      </span>
      <span>
        Option 1, Option 2
      </span>
    </li>
    <li>
      <span>
        Attribute 2
      </span>
      <span>
        Option 3
      </span>
    </li>
    <li>
      <span>
        Attribute 3
      </span>
    </li>
  </ul>
</div>
`;

exports[`#CustomAttributes renders without custom attributes 1`] = `null`;
